File: quests/chapter15/java/Distributed.java
// Chapter 15: Core of the Architect - Final Boss
// Distributed Systems: Consistent Hashing and CRDTs

import java.util.*;

/**
 * ConsistentHash implements consistent hashing for distributed systems.
 * Uses virtual nodes (replicas) to improve key distribution.
 */
class ConsistentHash {
    private TreeMap<Integer, String> ring;  // hash -> node mapping
    private List<String> nodes;              // list of physical nodes
    private int replicas;                    // number of virtual nodes per physical node

    public ConsistentHash(int replicas) {
        this.ring = new TreeMap<>();
        this.nodes = new ArrayList<>();
        this.replicas = replicas;
    }

    /**
     * Computes a hash for the given key.
     * Uses a simple but effective hash function.
     */
    private int hash(String key) {
        // TODO: Implement hash function
        // Use key.hashCode() and ensure positive value with Math.abs()
        // Or implement a custom hash like FNV-1a
        return 0;
    }

    /**
     * Adds a physical node to the hash ring.
     * Creates 'replicas' virtual nodes for better distribution.
     *
     * TODO: Implement adding a node to the consistent hash ring
     * 1. Check if node already exists in nodes list
     * 2. Add node to nodes list
     * 3. For each replica (0 to replicas-1):
     *    - Create virtual node key: node + "-" + i
     *    - Compute hash using hash()
     *    - Add to ring mapping hash -> node
     * 4. TreeMap automatically maintains sorted order
     */
    public void addNode(String node) {
        // TODO: Implement this method
    }

    /**
     * Removes a physical node and all its virtual nodes from the ring.
     *
     * TODO: Implement removing a node from the consistent hash ring
     * 1. Check if node exists and remove from nodes list
     * 2. For each replica (0 to replicas-1):
     *    - Create virtual node key: node + "-" + i
     *    - Compute hash and remove from ring
     */
    public void removeNode(String node) {
        // TODO: Implement this method
    }

    /**
     * Finds the node responsible for a given key.
     * Uses clockwise traversal on the hash ring.
     *
     * TODO: Implement getting the node for a key
     * 1. If ring is empty, return null
     * 2. Compute hash of the key
     * 3. Use ring.ceilingEntry(hash) to find first hash >= key hash
     * 4. If no entry found (key hash > all in ring), wrap around to first
     *    Use ring.firstEntry()
     * 5. Return the node from the entry
     */
    public String getNode(String key) {
        // TODO: Implement this method
        return null;
    }

    /**
     * Returns the list of all physical nodes.
     */
    public List<String> getNodes() {
        return new ArrayList<>(nodes);
    }

    /**
     * Returns the number of physical nodes in the ring.
     */
    public int size() {
        return nodes.size();
    }
}

// ============================================================================
// GCounter - A CRDT (Conflict-free Replicated Data Type) grow-only counter
// Each node maintains its own counter, and counters can only be incremented
// Merging takes the maximum value for each node
// ============================================================================

/**
 * GCounter represents a grow-only counter CRDT.
 */
class GCounter {
    private String nodeId;              // ID of the local node
    private Map<String, Long> counts;   // node ID -> count mapping

    public GCounter(String nodeId) {
        this.nodeId = nodeId;
        this.counts = new HashMap<>();
    }

    /**
     * Increments the counter for the local node by 1.
     *
     * TODO: Implement incrementing the local counter
     * Increment counts.get(nodeId) by 1 (handle null case)
     */
    public void increment() {
        // TODO: Implement this method
    }

    /**
     * Increases the counter for the local node by a specified amount.
     * Only allows positive increments (GCounter can only grow).
     *
     * TODO: Implement incrementing by a specific amount
     * Only allow positive increments
     * Increment counts.get(nodeId) by amount
     */
    public void incrementBy(long amount) {
        // TODO: Implement this method
    }

    /**
     * Returns the total count across all nodes.
     *
     * TODO: Implement getting the total value
     * Sum all values in counts map
     */
    public long value() {
        // TODO: Implement this method
        return 0;
    }

    /**
     * Returns the count for the local node only.
     */
    public long localValue() {
        return counts.getOrDefault(nodeId, 0L);
    }

    /**
     * Merges another GCounter into this one.
     * Takes the maximum value for each node (ensures convergence).
     *
     * TODO: Implement merging two GCounters
     * For each node in other.counts:
     *   counts.put(node, max(counts.get(node), other.counts.get(node)))
     * This ensures CRDTs converge to the same value
     */
    public void merge(GCounter other) {
        // TODO: Implement this method
    }

    /**
     * Returns a copy of the internal counts map.
     */
    public Map<String, Long> getCounts() {
        return new HashMap<>(counts);
    }

    /**
     * Returns the node ID.
     */
    public String getNodeId() {
        return nodeId;
    }
}

// ============================================================================
// PNCounter - A CRDT counter that supports both increment and decrement
// Implemented as two GCounters: one for increments (P) and one for decrements (N)
// ============================================================================

/**
 * PNCounter represents a positive-negative counter CRDT.
 */
class PNCounter {
    private String nodeId;    // ID of the local node
    private GCounter p;       // positive counter (increments)
    private GCounter n;       // negative counter (decrements)

    public PNCounter(String nodeId) {
        this.nodeId = nodeId;
        this.p = new GCounter(nodeId);
        this.n = new GCounter(nodeId);
    }

    /**
     * Increases the counter by 1.
     *
     * TODO: Implement increment using the P counter
     */
    public void increment() {
        // TODO: Implement this method
    }

    /**
     * Decreases the counter by 1.
     *
     * TODO: Implement decrement using the N counter
     */
    public void decrement() {
        // TODO: Implement this method
    }

    /**
     * Returns the current value (P - N).
     *
     * TODO: Implement getting the value
     * Return p.value() - n.value()
     */
    public long value() {
        // TODO: Implement this method
        return 0;
    }

    /**
     * Merges another PNCounter into this one.
     *
     * TODO: Implement merging two PNCounters
     * Merge both P and N counters separately
     */
    public void merge(PNCounter other) {
        // TODO: Implement this method
    }

    /**
     * Returns the positive counter for testing.
     */
    public GCounter getPositive() {
        return p;
    }

    /**
     * Returns the negative counter for testing.
     */
    public GCounter getNegative() {
        return n;
    }
}

// Main class for the stub file
public class Distributed {
    public static void main(String[] args) {
        System.out.println("Chapter 15: Core of the Architect - Final Boss");
        System.out.println("Distributed Systems: Consistent Hashing and CRDTs");
        System.out.println("\nImplement the TODOs in ConsistentHash, GCounter, and PNCounter classes!");
    }
}

File: quests/chapter15/rust/distributed.rs
// Chapter 15: Core of the Architect - Final Boss
// Distributed Systems: Consistent Hashing and CRDTs

use std::collections::{BTreeMap, HashMap};
use std::hash::{Hash, Hasher};
use std::collections::hash_map::DefaultHasher;

/// ConsistentHash implements consistent hashing for distributed systems
/// Uses virtual nodes (replicas) to improve key distribution
pub struct ConsistentHash {
    ring: BTreeMap<u64, String>,  // hash -> node mapping (sorted for efficient lookup)
    nodes: Vec<String>,           // list of physical nodes
    replicas: usize,              // number of virtual nodes per physical node
}

impl ConsistentHash {
    /// Creates a new consistent hash ring with the given number of replicas
    pub fn new(replicas: usize) -> Self {
        ConsistentHash {
            ring: BTreeMap::new(),
            nodes: Vec::new(),
            replicas,
        }
    }

    /// Computes a hash for the given key
    fn hash_key(key: &str) -> u64 {
        let mut hasher = DefaultHasher::new();
        key.hash(&mut hasher);
        hasher.finish()
    }

    /// Adds a physical node to the hash ring
    /// Creates 'replicas' virtual nodes for better distribution
    pub fn add_node(&mut self, node: &str) {
        // TODO: Implement adding a node to the consistent hash ring
        // 1. Check if node already exists in self.nodes
        // 2. Add node to self.nodes
        // 3. For each replica (0 to self.replicas):
        //    - Create virtual node key: format!("{}-{}", node, i)
        //    - Compute hash using Self::hash_key()
        //    - Insert into self.ring: hash -> node.to_string()
    }

    /// Removes a physical node and all its virtual nodes from the ring
    pub fn remove_node(&mut self, node: &str) {
        // TODO: Implement removing a node from the consistent hash ring
        // 1. Find and remove node from self.nodes (use retain or position+remove)
        // 2. For each replica (0 to self.replicas):
        //    - Create virtual node key: format!("{}-{}", node, i)
        //    - Compute hash and remove from self.ring
    }

    /// Finds the node responsible for a given key
    /// Uses clockwise traversal on the hash ring
    pub fn get_node(&self, key: &str) -> Option<&str> {
        // TODO: Implement getting the node for a key
        // 1. If ring is empty, return None
        // 2. Compute hash of the key using Self::hash_key()
        // 3. Use BTreeMap's range() to find the first hash >= key hash
        // 4. If no hash >= key hash found, wrap around to first node
        //    (Hint: use self.ring.iter().next())
        // 5. Return Some(node) with the node associated with that hash
        None
    }

    /// Returns a copy of all physical nodes
    pub fn get_nodes(&self) -> Vec<String> {
        self.nodes.clone()
    }

    /// Returns the number of physical nodes in the ring
    pub fn size(&self) -> usize {
        self.nodes.len()
    }
}

// ============================================================================
// GCounter - A CRDT (Conflict-free Replicated Data Type) grow-only counter
// Each node maintains its own counter, and counters can only be incremented
// Merging takes the maximum value for each node
// ============================================================================

/// GCounter represents a grow-only counter CRDT
pub struct GCounter {
    node_id: String,              // ID of the local node
    counts: HashMap<String, i64>, // node ID -> count mapping
}

impl GCounter {
    /// Creates a new GCounter for the given node
    pub fn new(node_id: &str) -> Self {
        GCounter {
            node_id: node_id.to_string(),
            counts: HashMap::new(),
        }
    }

    /// Increments the counter for the local node by 1
    pub fn increment(&mut self) {
        // TODO: Implement incrementing the local counter
        // Use entry().or_insert(0) and increment
        // *self.counts.entry(self.node_id.clone()).or_insert(0) += 1;
    }

    /// Increments the counter for the local node by a specified amount
    pub fn increment_by(&mut self, amount: i64) {
        // TODO: Implement incrementing by a specific amount
        // Only allow positive increments (GCounter can only grow)
        // if amount > 0 { ... }
    }

    /// Returns the total count across all nodes
    pub fn value(&self) -> i64 {
        // TODO: Implement getting the total value
        // Sum all values in self.counts: self.counts.values().sum()
        0
    }

    /// Returns the count for the local node only
    pub fn local_value(&self) -> i64 {
        // TODO: Return the local node's count
        // *self.counts.get(&self.node_id).unwrap_or(&0)
        0
    }

    /// Merges another GCounter into this one
    /// Takes the maximum value for each node (ensures convergence)
    pub fn merge(&mut self, other: &GCounter) {
        // TODO: Implement merging two GCounters
        // For each (node_id, count) in other.counts:
        //   let current = self.counts.entry(node_id.clone()).or_insert(0);
        //   *current = (*current).max(*count);
    }

    /// Returns a copy of the internal counts map
    pub fn get_counts(&self) -> HashMap<String, i64> {
        self.counts.clone()
    }
}

// ============================================================================
// PNCounter - A CRDT counter that supports both increment and decrement
// Implemented as two GCounters: one for increments (P) and one for decrements (N)
// ============================================================================

/// PNCounter represents a positive-negative counter CRDT
pub struct PNCounter {
    node_id: String,  // ID of the local node
    p: GCounter,      // positive counter (increments)
    n: GCounter,      // negative counter (decrements)
}

impl PNCounter {
    /// Creates a new PNCounter for the given node
    pub fn new(node_id: &str) -> Self {
        PNCounter {
            node_id: node_id.to_string(),
            p: GCounter::new(node_id),
            n: GCounter::new(node_id),
        }
    }

    /// Increments the counter by 1
    pub fn increment(&mut self) {
        // TODO: Implement increment using the P counter
        // self.p.increment();
    }

    /// Decrements the counter by 1
    pub fn decrement(&mut self) {
        // TODO: Implement decrement using the N counter
        // self.n.increment();
    }

    /// Returns the current value (P - N)
    pub fn value(&self) -> i64 {
        // TODO: Implement getting the value
        // self.p.value() - self.n.value()
        0
    }

    /// Merges another PNCounter into this one
    pub fn merge(&mut self, other: &PNCounter) {
        // TODO: Implement merging two PNCounters
        // self.p.merge(&other.p);
        // self.n.merge(&other.n);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_consistent_hash_basic() {
        let mut ch = ConsistentHash::new(3);
        ch.add_node("server-1");
        ch.add_node("server-2");
        assert_eq!(ch.size(), 2);
    }

    #[test]
    fn test_consistent_hash_get_node() {
        let mut ch = ConsistentHash::new(3);
        ch.add_node("server-1");
        ch.add_node("server-2");

        let node = ch.get_node("some-key");
        assert!(node.is_some());
    }

    #[test]
    fn test_gcounter_basic() {
        let mut gc = GCounter::new("node-1");
        gc.increment();
        gc.increment();
        assert_eq!(gc.value(), 2);
    }

    #[test]
    fn test_gcounter_merge() {
        let mut gc1 = GCounter::new("node-1");
        gc1.increment();
        gc1.increment();

        let mut gc2 = GCounter::new("node-2");
        gc2.increment();

        gc1.merge(&gc2);
        assert_eq!(gc1.value(), 3);
    }

    #[test]
    fn test_pncounter() {
        let mut pn = PNCounter::new("node-1");
        pn.increment();
        pn.increment();
        pn.decrement();
        assert_eq!(pn.value(), 1);
    }
}
